_posts/2019-04-24-lecture-27.md

```diff
@@ -0,0 +1,434 @@
+
+~~~
+for t from 1 to T:
+  doSomething();
+  updateParameters();
+  doSomethingElse();
+~~~
+
+Where the `updateParameters()` function looks something like:
+
+$$
+\theta^{t+1} = g(\theta^t, \Delta_f\theta(\mathcal{D}))
+$$
+
+where $\Delta_f\theta(\mathcal{D})$ is an incremental update calculated as a function of the dataset and the previous value of the parameters. Since function scales with both the data and the parameters, it is typically the most computationally expensive portion of the entire algorithm and benefits from parallelization.
+
+We can broadly categorize the parallelization techniques that we use into two categories:
+
+- **Data parallel methods:** The data is divided into smaller chunks and sent to different machines, which calculate updates relevant to a particular chunk, after which some sort of an aggregation step is performed to combine the work done by different machines.
+- **Model parallel methods:** The model parameters are divided into different sets and sent to different machines and each machine calculates the updates for those specific parameters.
+
+Of course, a combination of these two approaches is also possible.
+
+## Parallelization for Optimization Algorithms
+
+### A concrete optimization problem: Linear Regression
+
+Consider the case of a simple linear regression with a squared error loss. We wish to find:
+
+$$
+\beta^* = \arg\min_\beta \frac{1}{2}\lVert Y - \beta X\rVert_2^2 + \lambda\Omega(\beta)
+$$
+
+Where $\Omega$ is a regularization term which may be used to induce sparsity, or incorporate domain knowledge. The form of $\Omega$ may vary depending on what the underlying priors are about the structure of the problem. This tells us that linear regression is actually a collection of models, where we can pick a data-fitting loss function and a regularization term and take a weighted sum of the two. Therefore, any parallel algorithms that one might come up with should be fairly general.
+
+### Recap: The sequential Stochastic Gradient Descent (SGD)
+  Consider the following optimization problem:
+
+  $$
+  \theta^* = \arg\min_\theta \mathbb{E}_{x}[\mathcal{L(x;\theta)}]
+  $$
+
+  Vanilla gradient descent updates the parameters in the following manner:
+
+  $$
+  \theta^{t+1} = \theta^{t} - \gamma\frac{1}{n}\sum_{i=1}^{n}\nabla_{\theta^t}\mathcal{L(x_i;\theta^t)}
+  $$
+
+  Stochastic gradient descent uses a noisy estimate of the gradient using a single point:
+
+  $$
+  \theta^{t+1} = \theta^{t} - \gamma\nabla_{\theta^t}\mathcal{L(x_i;\theta^t)}
+  $$
+
+  This converges almost surely to the global optimum for convex problems. In practice, we actually use a batch of data points instead of a single one, which decreases the variance of the gradient and also allows us to use efficient matrix multiplication libraries.
+
+### Parallel Stochastic Gradient Descent (PSGD) <d-cite key="zinkevich2010parallelized"></d-cite>:
```

Comment:
Please format this section in paragraph form so that it reads more like a blog post, rather than bullet points from lecture slides. Here's an example that I liked: https://sailinglab.github.io/pgm-spring-2019/notes/lecture-13/